draco/tests/test_hard.py

```diff
@@ -1,90 +1,324 @@
+            """
     attribute((mark,type),m1,text).
     entity(encoding,m1,e1).
     attribute((encoding,channel),e1,x).
     entity(encoding,m1,e2).
     attribute((encoding,channel),e2,y).
 
-    :- violation(_).
     """
+        )
+        == ["encoding_no_field_and_not_count", "text_mark_without_text_channel"]
     )
 
-    assert not is_satisfiable(
-        p
-        + """
+    assert (
+        get_violations(
+            """
     attribute((mark,type),m1,text).
     entity(encoding,m1,e1).
     attribute((encoding,channel),e1,x).
     entity(encoding,m1,e2).
     attribute((encoding,channel),e2,y).
     attribute((encoding,channel),e3,text).
 
-    :- violation(_).
     """
+        )
+        == ["encoding_no_field_and_not_count", "text_mark_without_text_channel"]
     )
 
 
 def test_text_channel_without_text_mark():
-    b = hard.blocks["text_channel_without_text_mark"]
-    assert isinstance(b, Block)
-    p = helpers.program + b.program
 
-    assert is_satisfiable(
-        p
-        + """
+    assert (
+        get_violations(
+            """
     attribute((mark,type),m1,text).
     entity(encoding,m1,e1).
     attribute((encoding,channel),e1,text).
 
-    :- violation(_).
     """
+        )
+        == ["encoding_no_field_and_not_count"]
     )
 
-    assert not is_satisfiable(
-        p
-        + """
+    assert (
+        get_violations(
+            """
     attribute((mark,type),m1,bar).
     entity(encoding,m1,e1).
     attribute((encoding,channel),e1,text).
 
-    :- violation(_).
     """
+        )
+        == ["encoding_no_field_and_not_count", "text_channel_without_text_mark"]
+    )
+
+
+def test_bin_and_aggregate():
+
+    assert (
+        get_violations(
+            """
+    entity(encoding,root,e1).
+    attribute((encoding,channel),e1,x).
+    attribute((encoding,binning),e1,10).
+    entity(encoding,root,e2).
+    attribute((encoding,channel),e2,y).
+    attribute((encoding,aggregate),e2,count).
+
+    """
+        )
+        == ["encoding_no_field_and_not_count"]
+    )
+
+    assert (
+        get_violations(
+            """
+    entity(encoding,root,e1).
+    attribute((encoding,channel),e1,x).
+    attribute((encoding,binning),e1,10).
+    attribute((encoding,aggregate),e1,count).
+
+    """
+        )
+        == ["bin_and_aggregate"]
+    )
+
+
+def test_no_encodings():
+
+    assert (
+        get_violations(
+            """
+    entity(mark,root,1).
+    entity(encoding,1,2).
+
+    """
+        )
+        == ["encoding_no_field_and_not_count"]
+    )
+
+    assert (
+        get_violations(
+            """
+    entity(mark,root,1).
+    entity(encoding,1,2).
+    entity(mark,root,3).
+
+    """
+        )
+        == ["encoding_no_field_and_not_count", "no_encodings"]
+    )
+
+
+def test_repeat_channel():
+
+    assert (
+        get_violations(
+            """
+    attribute((encoding,channel),e0,x).
+    attribute((encoding,channel),e1,y).
+
+    """
+        )
+        == []
```

Comment:
this can still be `is_satisfiable` as that's equivalent to checking for no violations but faster